crates/miden-lib/asm/kernels/transaction/lib/memory.masm

```diff
@@ -687,33 +715,81 @@ end
 #! Outputs: [ptr]
 #!
 #! Where:
-#! - ptr is the memory address at which the account data begins.
+#! - ptr is the memory address at which the data of the currently used account begins.
 export.get_current_account_data_ptr
-    push.CURRENT_ACCOUNT_DATA_PTR mem_load
+    push.ACCOUNT_STACK_TOP_PTR mem_load
+    # => [account_stack_top_ptr]
+
+    mem_load
+    # => [current_account_data_ptr]
 end
 
-#! Sets the memory pointer of the current account data to the provided value.
+#! Adds the pointer to the account stack.
 #!
-#! Inputs:  [curr_account_data_ptr]
-#! Outputs: []
+#! It is done by writing the pointer to the element next to the stack top and moving the top pointer
+#! to it.
+#!
+#! Inputs:  [curr_account_data_ptr, foreign_account_id_prefix, foreign_account_id_suffix]
+#! Outputs: [foreign_account_id_prefix, foreign_account_id_suffix]
 #!
 #! Where:
-#! - curr_account_data_ptr is the memory pointer to the currently accessing account data.
-export.set_current_account_data_ptr
-    push.CURRENT_ACCOUNT_DATA_PTR mem_store
+#! - curr_account_data_ptr is the memory address
+#! - foreign_account_id_{prefix,suffix} are the prefix and suffix felts of the ID of the foreign
+#!   account whose procedure is going to be executed.
+#!
+#! Panics if:
+#! - the account stack is full, containing 64 accounts total.
+#! - the provided foreign account ID is equal to the native account ID.
+export.push_ptr_to_account_stack
+    # check that the account stack is not full
+    push.ACCOUNT_STACK_TOP_PTR mem_load dup push.MAX_ACCOUNT_STACK_PTR
+    lt assert.err=ERR_ACCOUNT_STACK_OVERFLOW
+    # => [account_stack_top_ptr, curr_account_data_ptr, foreign_account_id_prefix, foreign_account_id_suffix]
+
+    # check that foreign account ID is not equal to the native account ID
+    dup.3 dup.3 exec.get_native_account_id exec.account::is_id_equal
+    assertz.err=ERR_FOREIGN_ACCOUNT_CONTEXT_AGAINST_NATIVE_ACCOUNT
+    # => [account_stack_top_ptr, curr_account_data_ptr, foreign_account_id_prefix, foreign_account_id_suffix]
```

Comment:
This may work, but I was actually thinking checking that `curr_account_data_ptr != native_account_data_ptr`. should be a simpler check and could be a bit more robust in this context (i.e., right now we are trusting the caller that `curr_account_data_ptr` is not a pointer to the native account). Or would this not work for some reason?